plex/player.py

```diff
@@ -0,0 +1,186 @@
+from plex.server import Server
+
+from lib.utils import log, mediaProvider2str
+
+REPORTING_INTERVAL = 5 # seconds
+
+class Player(xbmc.Player):
+    def __init__(self):
+        '''Initializes the player'''
+        super(xbmc.Player, self).__init__()
+
+        self._providers = {}
+        self._monitor = xbmc.Monitor()
+        self._lock = threading.Lock()
+
+    def AddProvider(self, mediaProvider):
+        '''Adds a media provider to the player'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        self._providers[mediaProvider.getIdentifier()] = mediaProvider
+
+    def RemoveProvider(self, mediaProvider):
+        '''Removes the associated media provider'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        del self._providers[mediaProvider.getIdentifier()]
+
+    def onPlayBackStarted(self):
+        '''Triggered when xbmc.Player is started'''
+        self._reset()
+        self._file = self.getPlayingFile()
+
+    def onAVStarted(self):
+        '''Triggered when playback actually starts'''
+        self._startPlayback()
+        self._syncPlaybackState(time=0, state=PLEX_PLAYER_PLAYING)
+
+    def onPlayBackSeek(self, time, seekOffset):
+        '''Triggered when seeking.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackSeekChapter(self, chapter):
+        '''Triggered when seeking chapters.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackPaused(self):
+        '''Triggered when playback is paused.'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PAUSED)
+
+    def onPlayBackResumed(self):
+        '''Triggered when playback is resumed after a pause'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PLAYING)
+
+    def onPlayBackStopped(self):
+        '''Triggered when playback is stopped'''
+        self._syncPlaybackState(state=PLEX_PLAYER_STOPPED)
+        self.onPlayBackEnded()
+
+    def onPlayBackEnded(self):
+        '''Triggered when playback ends. Resets the player state and inherently kills the reporting loop'''
+        self._reset()
+
+    def _startPlayback(self):
+        '''Identifies the item (if from Plex) and initializes the player state'''
+        if not self._file:
+            return
+
+        if not self.isPlayingVideo():
+            return
+
+        videoInfoTag = self.getVideoInfoTag()
+        if not videoInfoTag:
+            return
+
+        itemId = videoInfoTag.getUniqueID(PLEX_PROTOCOL)
+        if not itemId:
+            return
+
+        if not itemId.isdigit():
+            log('invalid item id {} (non digit). Kodi will not report playback state to Plex Media Server' \
+                .format(matchingItems[0]), xbmc.LOGERROR)
+
+        self._itemId = int(itemId)
+
+        for mediaProvider in self._providers.values():
```

Comment:
I don't really like this approach as it makes you iterate over all the imported items from every provider. Isn't there an alternative implementation  to get it directly from the mediaimport API?

Reply:
How would such a call look like? You only have a file or a list item. The fact that the identification happens over the unique ID is specific to the media importer implementation. Another media importer could use a different approach (e.g. determine the unique identifier from the file name). Furthermore you have no idea from which media provider / import the item is coming from.

I'd have to check if we could make the information from which media provider an item has been imported available on the list item. Then you would only have to retrieve all imported items from the media provider matching the item's assign media provider.

Reply:
Hum can you set both the media provider id and the media importer id as listitem properties? That way we could from python just check if both properties are defined and later query the mediaimport API for the specific provider/importer objects providing the Id

Reply:
I could provide APIs to get the media provider identifier and the media import path through the `ListItem`'s class. Then you can simply use `self._providers[item.getMediaProviderId()]`.

Reply:
I just realized that the `xbmc.Player` API does not give us access to the `ListItem` so that would have to be added as well. I've created an improvement at https://github.com/Montellese/xbmc/issues/60 but IMO we should not block this PR due to this. Maybe add a TODO comment that this should be improved.

Reply:
That would be awesome since there's a pending issue in Kodi that to be solved needs the Listitem to be accessed from the player (json-rpc). I tried my luck with that issue but it seems that it has to be implemented in ApplicationPlayer and then proxied to every single player (as far I as could tell). I failed miserably doing it for retroplayer and UPnPPlayer (they don't hold a listitem). PAPlayer and VideoPlayer seems to be easy.

I'll get back to this PR as soon as possible

Reply:
I've fixed this in 70ee52452b9382d6f63fb492651cf4e3f3354b22 after I added `Player.getPlayingItem()` and `ListItem.getMediaProviderId()` to Kodi's python API. That way we can directly lookup the matching media provider from `self._providers` using the item's media provider ID.